### PR TITLE
feat: Introduce p2p slot table limit and limit allowed ingress slots per peer

### PR DESCRIPTION
**Pull request by tthebst**
To support hashes in blocks there are modifications #1061 to the ingress pool that would allow the unvalidated ingress pool to grow without bounds. P2P allows for bounding the unvalidated pool via the slot table. Until now the slot table didn't have any bounds and this MR introduces a config option to bound the slot table and introduces a hard bound for the ingress pool of `50000`. Other clients use `usize::MAX` as a bound.

### CODE REVIEW

File: rs/p2p/consensus_manager/src/receiver.rs [hunk 1]
```diff
@@ -364,14 +365,25 @@ where
...
             }
+            Entry::Vacant(_) => {
+                self.metrics.slot_table_limit_exceeded_total.inc();
+                warn!(
+                    self.log,
+                    "Peer {} tries to exceed slot limit {}. Dropping slot update",
+                    peer_id,
+                    self.slot_limit
+                );
+                return;
```
Comment (kpop-dfinity) [inline rs/p2p/consensus_manager/src/receiver.rs:385]: tiny nit: to be consistent with a branch above (line 365)
```suggestion
                (false, None)
```

**Review by kpop-dfinity — APPROVED** [review]
Thanks for speedy implementation!

**Comment by rumenov** [comment]
Approved but please explain this is only for malicious peers and that we should have a bound on the validated pool by originator in order to bound the unvalidated pool in the happy case.

**Review by rumenov — APPROVED** [review]

**Comment by DSharifi** [comment]
Can we have an edge case where:
1. A  sender reaches the max pool size, and is transmitting all these artifacts on active slots.
2.  Its next batch of artifacts includes the same number of removals and additions.
3. P2P processes the additions first, meaning we go over the slot size before we prune the slot table with removals.

**Review by DSharifi — APPROVED** [review]

**Comment by tthebst** [comment]
Good point about the edge case. I think this is not possible anymore since the interface is now `Vec<Transmits>` where `Transmits` is either `Deliver` or `Abort`. So consensus defines the ordering of additions and deletions and we assume consensus will make sure that we will stay within the limit. 

Also note that the limit is for now only applied to ingress.

**Pull request merged**